keps/sig-node/20190221-maxInitialFailureCount-health-probes.md

```diff
@@ -0,0 +1,145 @@
+- The container fails 10 probes at startup, starts successfully, and after a long time fails 3 probes. The container is considered to have failed and is terminated after 30 seconds of downtime.
+- The container fails 10 probes at startup, succeeds once, and fails 3 more probes. The container is considered to have started at 100 seconds, and even though it is still within the first 200 seconds of its lifetime covered by `maxInitialFailureCount`, it is considered to have failed because of `failureThreshold` and the fact that it had initially started successfully, therefore it is terminated after 30 seconds of downtime.
+
+You can see the implementation in PR [#71449].
+
+[#71449]: https://github.com/kubernetes/kubernetes/pull/71449
+
+### Risks and Mitigations
+
+#### Stateless kubelet
+
+`kubelet` handles container/pod lifecycle-related functions by relying on the underlying container runtime to persist the states. For probes and/or lifecycle hooks, kubelet rely on in-memory states only.
+
+This means that the boolean `hasInitialized` as well as all probe counters could be reset during the lifetime of a container. There are several cases to consider:
+
+- The container is starting: `hasInitialized` was still false, which means the container will have more probe attempts to successfully start. This is also the case today with a long `initialDelaySeconds`, except that with the new feature `failureThreshold` is taken into account after the first success of the probe.
+- The container is running: `hasInitialized` is reverted to false until the next probe which will succeed and reset it to true immediately after.
+- The container is dead-locked: `hasInitialized` is reverted to false, which means the container will have a maximum downtime of `periodSeconds` times `maxInitialFailureCount`. This is also the case today with a high `failureThreshold`.
+
+#### maxInitialFailureCount is set smaller than failureThreshold
+
+Such a configuration could lead to the unexpected behaviour of having the container being killed faster before startup than it would in case of a dead-lock.
+
+A reasonable failsafe value would be to set `maxInitialFailureCount` to `failureThreshold` when it happens.
+
+## Design Details
+
+### Test Plan
+
+The following test cases can be covered by calling the `fakeExecProber` a number of times to verify:
+
+- the container is killed after `maxInitialFailureCount` if it has never initialized (emulated by always calling `fakeExecProber{probe.Success, nil}`)
+- the container is killed after `failureThreshold` once it has initialized (emulated by calling `fakeExecProber{probe.Success, nil}` once) with total probes > `maxInitialFailureCount`
+- the container is killed after `failureThreshold` once it has initialized (emulated by calling `fakeExecProber{probe.Success, nil}` once) with total probes < `maxInitialFailureCount`
+
+## Implementation History
```

Comment:
You've also implemented (and tested?) code in the PR, correct?  You should note the prototype that is concurrently under review.

Reply:
yes, works well in my case - this is why I am pushing for this feature